design/secondary-network-ipam.md

```diff
@@ -0,0 +1,181 @@
+network in some deployments, get confused if a client IP changes. We
+therefore want to ensure that those IPs do not change.
+
+We use dnsmasq to manage PXE booting servers during
+provisioning. dnsmasq will not bind to an interface managed by
+dhclient, so at least some of the hosts must have statically allocated
+IPs on the secondary network to allow us to run dnsmasq at all. This
+also means it is not sufficient to manage DHCP reservations to ensure
+a given host always receives the same IP.
+
+When we implement host discovery, we will want to allow discovered
+hosts to use part of the IP range on the provisioning network that is
+not used for static allocations so that a user does not have to clean
+up those static allocations for hosts they are not using in their
+cluster.
+
+To meet all of these requirements, we need to configure the secondary
+network interfaces on each host with a static IP address.
+
+### Risks and Mitigations
+
+We need to ensure the DHCP address range and static address range do
+not overlap. We should be able to ensure that with careful management
+of the CIDRs.
+
+[inwinstack/ipam](https://github.com/inwinstack/ipam) may not be
+stable or reliable, and we would have to either fix it, fork it, or
+build a replacement.
+
+## Design Details
+
+We need to divide the subnet range for the provisioning network
+between a set of addresses we can use for DHCP and a set for static
+IPs.
+
+We need the installer to allocate IPs for the master nodes as it
+provisions them, and to record that information in the kubernetes
+database so those same IPs are not used for other hosts later.
+
+We need to store the subnet CIDR and existing allocations in the
+kubernetes database somewhere so new IPs can be allocated when hosts
+are provisioned.
+
+The [inwinstack/ipam](https://github.com/inwinstack/ipam) controller
+provides `Pool` and `IP` resources for allocating IPs from address
+ranges. We should evaluate it to see if we can use it for managing the
+IP allocations.
+
+The machine-api-provider-baremetal controller is responsible for
+making decisions about how to configure a host, so it should request
+IPs for secondary networks, assign them to the interfaces, and pass
+the relevant data using the ignition configuration data. It will need
+to create host-specific ignition configuration resources because it
+will be different for each host. It should also set the `Machine` as
+an owner of the `IP` so that the reservation is deleted when the
+`Machine` is deleted.
+
+### Work Items
+
+1. Ensure the IP ranges for secondary networks are captured by the
+   installer and saved to the kubernetes database as `Pool` resources.
+1. Ensure the installer registers the IP allocations for masters.
+1. Ensure the IPAM service is deployed along with the other metal3
+   components.
+1. Update the metal3 machine controller to allocate IPs and create
+   host-specific ignition configurations containing the IPs.
```

Comment:
A few of these work items are OpenShift / CoreOS specific, the "installer" and "ignition" references.

The machine controller is not involved today in creating the ignition config (or cloud-init config, or whatever user data is in use), so maybe this should go somewhere else.

I expect management of secondary interfaces should be done by something else, like https://github.com/nmstate/kubernetes-nmstate

an IPAM component allocating addresses for secondary network interfaces could then create the CRs that specify that the interface should be configured with that IP, and applying the configuration would be done by kubernetes-nmstate

Reply:
Yeah, I was trying to make sure I didn't miss anything. Should I move this doc to an internal location?

Reply:
no need for it to be internal.  It just depends if we come up with something that's more generally useful, or is an OpenShift specific integration thing.  If it's just for OpenShift integration, the openshift-metal3 github org has a docs repo as well that hasn't been used much yet.

Speaking of OpenShift specific solutions and how to do network config ... I wonder if the existing MachineConfig resource is enough, provided by the machine-config-operator.  We can create those resources to drop new config files on nodes, or to replace existing config files.